add swtpm demo

DNJha opened the pull request:
This PR:
- adds software trusted platform module (SWTPM) to run inside SGX
- test script for IBM's TPM2.0 TSS

Tested on Azure confidential Computing Environment (_Standard DC2ds v3_) with Ubuntu 20.04.

jessehui commented:
@DNJha Thank you for your delicate work! This is a great contribution to the Occlum project. I took a quick look at this PR.  In general, this looks really good. There are just minor questions regarding this PR:
- If I understand correctly, the `demos/swtpm/run_swtpm.sh` only starts the server, but the client command is not included. Could you also add a client command to make this test script complete?
- With the complete script, could you also add this demo to our CI? Most of the demos have their CI defined at `.github/workflows/demo_test.yml`. This can help us make sure the demos work for new Occlum versions.
- Please also include SWTPM at demos/README.md with just a simple line

Thank you!

jessehui reviewed (commented):
Thank you for your delicate work! This is a great contribution to the Occlum project. I took a quick look at this PR. In general, this looks really good. There are just minor questions regarding this PR:

- If I understand correctly, the demos/swtpm/run_swtpm.sh only starts the server, but the client command is not included. Could you also add a client command to make this test script complete?

- With the complete script, could you also add this demo to our CI? Most of the demos have their CI defined at .github/workflows/demo_test.yml. This can help us make sure the demos work for new Occlum versions.

- Please also include SWTPM at demos/README.md with just a simple line

Thank you!

DNJha commented:
As suggested, I added IBM's TPM2.0 TSS client to test the functioning of SWTPM, updated the README file and defined the script for CI.

jessehui commented:
@DNJha Thank you for updating! But it looks like the CI failed. I think it is due to the formatting issue of the `swtpm.yaml`. Please take a look. After fixing this, please squash all the commits into a single one. And you can force push to your branch of this PR. Thank you.

I have another question when running your demo. When running under my server, it looks like the `make check` command in the `install_swtpm.sh` failed. I checked the CI status and it looks like with GitHub's CI machine, all checks passed. Do you know why and how can I fix this?
<img width="899" alt="截屏2023-07-31 09 53 49" src="https://github.com/occlum/occlum/assets/13930406/be0be231-577f-481a-8f8b-8d63f114f91b">

DNJha commented:
@jessehui I have updated the `swtpm.yaml` file as well as the `install_swtpm.sh` to replace wget with a git clone. I also squashed the commits to one commit as suggested. 

The `make check ` was failing because of the update on `libtpms`. I added the dependencies to resolve that, I hope this will work now. Please let me know if it still fails.

jessehui commented on demos/swtpm/run_swtpm.sh:
```diff
@@ -0,0 +1,25 @@
+echo -e "${BLUE}Occlum start swtpm${NC}"
+
+occlum run /bin/swtpm socket --tpmstate dir=/bin/myvtpm --tpm2 --ctrl type=tcp,port=2322,bindaddr=0.0.0.0 --server type=tcp,port=2321,bindaddr=0.0.0.0 
+--flags not-need-init --seccomp action=none
```
This script can't run correctly. This command should be in one line.

DNJha replied:
Sorry for the typo, I fixed it now

jessehui commented on .github/workflows/demo_test.yml:
```diff
@@ -871,3 +871,24 @@ jobs:
+      run: docker exec ${{ github.job }} bash -c "cd /root/occlum/demos/swtpm; ./install_swtpm.sh"
+
+    - name: Run swtpm server
+      run: docker exec ${{ github.job }} bash -c "cd /root/occlum/demos/swtpm; SGX_MODE=SIM ./run_swtpm.sh"
```
`run_swtpm.sh` will run as a server and this script will never return. You can take a look at the `Https_server_test` test in this file to use the background process to run this test.

DNJha replied:
I updated it following the` Https_server_test`. Hope the CI will pass this time

jessehui commented:
@DNJha Thanks for your update. However, with the latest code, I still got many failure cases when running the `install_swtpm.sh` script.  Do you have any idea why this happened?
![截屏2023-08-23 17 00 44](https://github.com/occlum/occlum/assets/13930406/e70c8623-6c76-4940-91a5-292f3cb95083)

This check error doesn't prevent me from running this demo. I give two extra comments. Please take a look. Thank you!

jessehui reviewed (approved):
LGTM. Thank you for your contribution!

jessehui commented:
@DNJha Thank you for your update. I think it's fine. Without passing all the checks, this demo can also run. And it seems the CI has passed. I think it is close to merging.

DNJha commented:
Thank you.
